File: apple/OmnivoreKit/Sources/Models/Highlight.swift
import CoreData
import Foundation

public struct Highlight: Identifiable, Hashable, Codable {
  public let id: String
  public let shortId: String
  public let quote: String
  public let prefix: String?
  public let suffix: String?
  public let patch: String
  public let annotation: String?
  public let createdAt: Date?
  public let updatedAt: Date?
  public let createdByMe: Bool

  public init(
    id: String,
    shortId: String,
    quote: String,
    prefix: String?,
    suffix: String?,
    patch: String,
    annotation: String?,
    createdByMe: Bool,
    createdAt: Date? = nil,
    updatedAt: Date? = nil
  ) {
    self.id = id
    self.shortId = shortId
    self.quote = quote
    self.prefix = prefix
    self.suffix = suffix
    self.patch = patch
    self.annotation = annotation
    self.createdAt = createdAt
    self.updatedAt = updatedAt
    self.createdByMe = createdByMe
  }

  public func toManagedObject(context: NSManagedObjectContext, associatedItemID: String) -> PersistedHighlight {
    let persistedHighlight = PersistedHighlight(context: context)
    persistedHighlight.associatedItemId = associatedItemID
    persistedHighlight.markedForDeletion = false
    persistedHighlight.id = id
    persistedHighlight.shortId = shortId
    persistedHighlight.quote = quote
    persistedHighlight.prefix = prefix
    persistedHighlight.suffix = suffix
    persistedHighlight.patch = patch
    persistedHighlight.annotation = annotation
    persistedHighlight.createdAt = createdAt
    persistedHighlight.updatedAt = updatedAt
    persistedHighlight.createdByMe = createdByMe
    return persistedHighlight
  }

  public static func make(from persistedHighlight: PersistedHighlight) -> Highlight {
    Highlight(
      id: persistedHighlight.id ?? "",
      shortId: persistedHighlight.shortId ?? "",
      quote: persistedHighlight.quote ?? "",
      prefix: persistedHighlight.prefix,
      suffix: persistedHighlight.suffix,
      patch: persistedHighlight.patch ?? "",
      annotation: persistedHighlight.annotation,
      createdByMe: persistedHighlight.createdByMe,
      createdAt: persistedHighlight.createdAt,
      updatedAt: persistedHighlight.updatedAt
    )
  }
}

File: android/Omnivore/app/src/main/java/app/omnivore/omnivore/core/database/entities/Highlight.kt
package app.omnivore.omnivore.core.database.entities

import androidx.room.Dao
import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Insert
import androidx.room.Junction
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Relation
import app.omnivore.omnivore.core.data.model.ServerSyncStatus
import com.google.gson.annotations.SerializedName


@Entity
data class Highlight(
  @SerializedName("id")
  @PrimaryKey val highlightId: String,
  val type: String,
  var annotation: String?,
  val createdAt: String?,
  val createdByMe: Boolean = true,
  val markedForDeletion: Boolean = false,
  var patch: String?,
  var prefix: String?,
  var quote: String?,
  var serverSyncStatus: Int = ServerSyncStatus.IS_SYNCED.rawValue,
  var shortId: String,
  val suffix: String?,
  val updatedAt: String?,
  val color: String?,
  val highlightPositionPercent: Double?,
  val highlightPositionAnchorIndex: Int?
)

@Entity(
  primaryKeys = ["highlightId", "savedItemId"],
  foreignKeys = [
    ForeignKey(
      entity = Highlight::class,
      parentColumns = arrayOf("highlightId"),
      childColumns = arrayOf("highlightId"),
      onDelete = ForeignKey.CASCADE
    ),
    ForeignKey(
      entity = SavedItem::class,
      parentColumns = arrayOf("savedItemId"),
      childColumns = arrayOf("savedItemId"),
      onDelete = ForeignKey.CASCADE
    )
  ]
)
data class SavedItemAndHighlightCrossRef(
  val highlightId: String,
  val savedItemId: String
)

@Dao
interface SavedItemAndHighlightCrossRefDao {
  @Insert(onConflict = OnConflictStrategy.REPLACE)
  fun insertAll(items: List<SavedItemAndHighlightCrossRef>)

  @Query("SELECT savedItemId FROM savedItemAndHighlightCrossRef WHERE highlightId = :highlightId")
  fun associatedSavedItemID(highlightId: String): String?
}

data class SavedItemWithLabelsAndHighlights(
  @Embedded val savedItem: SavedItem,

  @Relation(
    parentColumn = "savedItemId",
    entityColumn = "savedItemLabelId",
    associateBy = Junction(SavedItemAndSavedItemLabelCrossRef::class)
  )
  val labels: List<SavedItemLabel>,

  @Relation(
    parentColumn = "savedItemId",
    entityColumn = "highlightId",
    associateBy = Junction(SavedItemAndHighlightCrossRef::class)
  )
  val highlights: List<Highlight>
) {
  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as SavedItemWithLabelsAndHighlights

    return savedItem.savedItemId == other.savedItem.savedItemId
  }

  override fun hashCode(): Int {
    return savedItem.savedItemId.hashCode()
  }
}
